tket/src/WeightSubgrMono/Searching/SearchBranch.cpp

```diff
@@ -0,0 +1,309 @@
+      m_derived_graphs_reducer(m_pattern_ndata, m_target_ndata),
+      m_neighbours_reducer(m_pattern_ndata, m_target_ndata),
+      m_nodes_raw_data_wrapper(initial_domains),
+      m_domains_accessor(m_nodes_raw_data_wrapper),
+      m_node_list_traversal(m_nodes_raw_data_wrapper) {
+  m_extra_statistics.number_of_pattern_vertices =
+      m_pattern_ndata.get_number_of_nonisolated_vertices();
+  m_extra_statistics.number_of_target_vertices =
+      m_target_ndata.get_number_of_nonisolated_vertices();
+  m_extra_statistics.initial_number_of_possible_assignments = 0;
+  for (const auto& entry : initial_domains) {
+    m_extra_statistics.initial_number_of_possible_assignments += entry.size();
+  }
+
+  // In what order should we do reduction/checks?
+  // The simplest/cheapest first? Most powerful?
+  // Seems a difficult question...
+  m_reducer_wrappers.emplace_back(m_neighbours_reducer);
+  m_reducer_wrappers.emplace_back(m_derived_graphs_reducer);
+
+  if (max_distance_reduction_value >= 2) {
+    m_distance_reducers.reserve(max_distance_reduction_value - 1);
+    m_reducer_wrappers.reserve(max_distance_reduction_value - 1);
+    for (unsigned distance = 2; distance <= max_distance_reduction_value;
+         ++distance) {
+      m_distance_reducers.emplace_back(
+          pattern_near_ndata, m_target_ndata, target_near_ndata, distance);
+    }
+    // Now that all the reducer objects are stored
+    // (the vector will not be resized),
+    // we can safely take references to them.
+    for (auto& reducer : m_distance_reducers) {
+      m_reducer_wrappers.emplace_back(reducer);
+    }
+  }
+}
+
+const DomainsAccessor& SearchBranch::get_domains_accessor() const {
+  return m_domains_accessor;
+}
+DomainsAccessor& SearchBranch::get_domains_accessor_nonconst() {
+  return m_domains_accessor;
+}
+
+std::set<VertexWSM> SearchBranch::get_used_target_vertices() const {
+  return m_node_list_traversal.get_used_target_vertices();
+}
+
+const ExtraStatistics& SearchBranch::get_updated_extra_statistics() {
+  if (m_weight_checker_ptr) {
+    const auto weight_checker_data_opt =
+        m_weight_checker_ptr->get_tv_data_opt();
+    if (weight_checker_data_opt) {
+      m_extra_statistics.n_tv_initially_passed_to_weight_nogood_detector =
+          weight_checker_data_opt->initial_number_of_tv;
+      m_extra_statistics.n_tv_still_valid_in_weight_nogood_detector =
+          weight_checker_data_opt->final_number_of_tv;
+    }
+  }
+  m_extra_statistics.total_number_of_assignments_tried = 0;
+  for (const auto& entry : m_checked_assignments) {
+    m_extra_statistics.total_number_of_assignments_tried += entry.second.size();
+  }
+  return m_extra_statistics;
+}
+
+void SearchBranch::activate_weight_checker(WeightWSM total_p_edge_weights) {
+  m_weight_checker_ptr = std::make_unique<WeightChecker>(
+      m_pattern_ndata, m_target_ndata, *this, total_p_edge_weights,
+      m_impossible_target_vertices);
+  TKET_ASSERT(m_weight_checker_ptr);
```

Comment:
Again, I was worried about unique ptr silently failing and returning null. However, it seems like that's not possible, something like std::bad_alloc will be thrown instead. (Maybe I was thinking of dynamic_cast or something else which CAN return null pointers?) I'm terrified of null ptr dereference (but I supposed it's not so bad nowadays, it probably won't crash your whole computer or overwrite your hard drive, unlike in the past...!)

Reply:
Please can ya comment here saying "done" or similar when it's removed

Reply:
Really rubbish, the docs for `make_unique` say "May throw `std::bad_alloc` or any exception thrown by the constructor of `T`".

Does this actually mean: "GUARANTEED to throw `std::bad_alloc` if memory allocation fails" ???

Surely that's what SHOULD should happen, but I want to be SURE! (Rather than it being really stupid and, say, returning a null ptr...)